diesel/src/sqlite/connection/raw.rs

```diff
@@ -194,6 +243,111 @@ extern "C" fn run_custom_function<F>(
+    value_ptr: *mut *mut ffi::sqlite3_value,
+) where
+    A: SqliteAggregateFunction<Args, Output = Ret> + 'static + Send,
+    Args: Queryable<ArgsSqlType, Sqlite>,
+    Ret: ToSql<RetSqlType, Sqlite>,
+    Sqlite: HasSqlType<RetSqlType>,
+{
+    unsafe {
+        let aggregate_context = ffi::sqlite3_aggregate_context(
+            ctx,
+            std::mem::size_of::<OptionalAggregator<A>>() as i32,
+        );
+        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+        let aggregator = match aggregate_context.map(|a| &mut *a.as_ptr()) {
+            Some(&mut OptionalAggregator::Some(ref mut agg)) => agg,
+            Some(mut a_ptr @ &mut OptionalAggregator::None) => {
+                ptr::write_unaligned(a_ptr as *mut _, OptionalAggregator::Some(A::default()));
+                if let &mut OptionalAggregator::Some(ref mut agg) = a_ptr {
+                    agg
+                } else {
+                    unreachable!(
+                        "We've written the aggregator above to that location, it must be there"
+                    )
+                }
+            }
+            None => {
+                null_aggregate_context_error(ctx);
+                return;
+            }
+        };
+
+        let mut f = |args: &[*mut ffi::sqlite3_value]| -> Result<(), Error> {
+            let args = build_sql_function_args::<ArgsSqlType, Args>(args)?;
+
+            Ok(aggregator.step(args))
+        };
+
+        let args = slice::from_raw_parts(value_ptr, num_args as _);
+        match f(args) {
+            Err(e) => {
+                let msg = e.to_string();
+                ffi::sqlite3_result_error(ctx, msg.as_ptr() as *const _, msg.len() as _);
+            }
+            _ => (),
+        };
+    }
+}
+
+extern "C" fn run_aggregator_final_function<ArgsSqlType, RetSqlType, Args, Ret, A>(
+    ctx: *mut ffi::sqlite3_context,
+) where
+    A: SqliteAggregateFunction<Args, Output = Ret> + 'static + Send,
+    Args: Queryable<ArgsSqlType, Sqlite>,
+    Ret: ToSql<RetSqlType, Sqlite>,
+    Sqlite: HasSqlType<RetSqlType>,
+{
+    unsafe {
+        // Within the xFinal callback, it is customary to set nBytes to 0 so no pointless memory
+        // allocations occur
+        // See: https://www.sqlite.org/c3ref/aggregate_context.html
+        let aggregate_context = ffi::sqlite3_aggregate_context(ctx, 0);
+        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+        let aggregator = match aggregate_context {
+            Some(ref mut a) => match std::mem::replace(a.as_mut(), OptionalAggregator::None) {
+                OptionalAggregator::Some(agg) => agg,
+                OptionalAggregator::None => unreachable!("We've written to the aggregator in the xStep callback. If xStep was never called, then ffi::sqlite_aggregate_context() would have returned a NULL pointer")
+            },
+            None => return,
+        };
```

Comment:
One last simplification/improvement:
This should remove the nested match statement and make it a bit clearer that a null value is returned by the none case, by explicitly setting the corresponding value, instead of implicitly doing that because of sqlite default behaviour:
```suggestion
        let aggregator = match aggregate_context.map(|a| std::mem::replace(a.as_mut(), OptionalAggregator::None)) {
            Some(OptionalAggregator::Some(agg)) => agg,
            Some(OptionalAggregator::None) => unreachable!("We've written to the aggregator in the xStep callback. If xStep was never called, then ffi::sqlite_aggregate_context() would have returned a NULL pointer"),
            None => {
                // explicitly set a null value as return value here, just to make it obvious what's returned in this case
                SerializedValue {
                    ty: Sqlite::metadata(&()),
                    data: None,
                }.result_of(ctx);
                return;
            },
        };
```